modules/custom/download_count/src/Plugin/Field/FieldFormatter/FieldDownloadCount.php

```diff
@@ -28,20 +125,46 @@ public function viewElements(FieldItemListInterface $items, $langcode) {
     $element = [];
     $entity = $items->getEntity();
     $entity_type = $entity->getEntityTypeId();
-    $access = \Drupal::currentUser()->hasPermission('view download counts');
+    $access = $this->user->hasPermission('view download counts');
 
     foreach ($this->getEntitiesToView($items, $langcode) as $delta => $file) {
       $item = $file->_referringItem;
 
       if ($access) {
-        $download = Database::getConnection()
-          ->query('SELECT COUNT(fid) from {download_count} where fid = :fid AND type = :type AND id = :id', [
-            ':fid' => $file->id(),
-            ':type' => $entity_type,
-            ':id' => $entity->id(),
-          ])
-          ->fetchField();
-        $file->download = $download;
+        // Try to fetch the count from cache.
+        $cache_data = $this->cache->get('download_count');
+
+        // Decode the cache json data.
+        $cache_download_count_array = $cache_data ? $this->serializer->decode($cache_data->data, 'json') : NULL;
+
+        // Get the count from cached array.
+        if ($cache_download_count_array && array_key_exists($file->id(), $cache_download_count_array)) {
+          $download_count = $cache_download_count_array[$file->id()];
+        }
+        else {
+          // If there is no data in cache, we find it via a query.
+          $download_count = $this->database->select('download_count', 'du')
+            ->fields('fu', ['fid'])
+            ->condition('type', $entity_type)
+            ->condition('id', $entity->id())
+            ->countQuery()
+            ->execute()
+            ->fetchField();
+
+          // Update the array.
+          $cache_download_counts[$file->id()] = $download_count;
+          // And then we set the data in cache again after encoding.
+          // This cache should automatically expire after an hour.
+          $this->cache->set('download_count', $this->serializer->encode($cache_download_counts, 'json'), time() + 3600);
```

Comment:
What you could do here instead of caching it for 1 hour, you could cache it indefinitely, whenever someone clicks on the download link you clear this cache and add 1 to the count and cache it again immediately for everybody.

Reply:
@ronaldtebrake Couldn't agree more.